### DISCUSSION TITLE
require() of ES Module not Supported from nested dependency

### DISCUSSION THREAD

**Discussion by timbomckay** [vitest-dev/vitest#3221] [Q&A] [answered]
We have an internal component library that's currently provided as CommonJS and is externalizing an ESM package via `require()`. They're working on providing their packages as modules but that'll take some time as they're trying to solve for SSR alongside this update.

I understand the error and that a CommonJS externalizing an ESM dependency is prone to these kinda errors, but I'm primarily confused as to why it's working with Vite but not Vitest. Our local development and bundling is working and figured it'd be the same for tests, but that doesn't appear to be the case.

<img width="866" alt="image" src="https://user-images.githubusercontent.com/784904/233408670-70b6d01b-5179-49a9-81bb-60007e097d2e.png">

- I've seen the various discussions and issue threads under this repo
- I've tried all variations of commonjs plugins
- I've tried inline, registerNodeLoader, and fallbackCJS
- I've tried transforming it to replace `require` with `import` but I got _"variable is not a function"_ errors afterwards

My assumption is we're gonna have to use something else until our component library is updated appropriately, just curious if there's something I'm missing for why Vite is behaving differently from Vitest.

**Comment by sheremet-va** [comment, marked as answer]
The behavior is different because Vite bundles your node_modules dependencies for the browser and Vitest doesn't. Vitest assumes you use dependencies that work in Node.js. There is `deps.experimentalOptimizer` field that enables bundling, but it doesn't work since 0.30.0 until we figure out the best way to resolve issues that appear.

**Reply by sheremet-va** [reply]
But I'm also not sure why would anyone bundle dependency in CJS, when it doesn't work (require is only supported in Node.js and Node.js doesn't allow requiring ESM). If it's meant to be consumed by the tools that support ESM, just bundle ESM and omit CJS.

**Reply by timbomckay** [reply]
Yeah providing ESM packages has been on the agenda for a couple years, but priorities, capacity, and consumer architecture have been factors. Most consuming applications were on Node 10 or 12 up until earlier this year.

Thanks for the details though, and the quick response. Nice to understand the differences.
